### PR TITLE
feat: group by endpoint

### LINKED ISSUE #448: Support Custom Batching
Sometimes batching support is provided by services directly for example in on **jsonplaceholder**

**GET One**
```curl
GET https://jsonplaceholder.typicode.com/users/1
```

**GET Many**
```curl
GET https://jsonplaceholder.typicode.com/users?id=1&id=2&id=3
```

When enabled this can reduce the load upstream drastically. We could make One Http call instead of many to load data for multiple user-ids.

**Approach**
- Create a DL (Data Loader) corresponding to each endpoint. This needs to be created once at the time of server start.
- All batchable requests (GET, DELETE etc.) should go via their corresponding Data Loader.
- Endpoint will need to maintain batching information which is how is the URL going to be generated if it's a call for many ids.
- DL should use the information in the endpoint about batching and perform additional operations if the endpoint supports batching.

**DSL Changes**
```graphql
type Query {
  posts: [Post] @http(path: "/posts")
  post(id: Int): Post @http(path: "/posts/:id")
}

type Post {
  title: String
  userId: User

  user: User 
    @http(path: "/users", query: [{key: "id", value: "{{value.userId}}"}]) 
    @batch(key: "userId", path: ["id"])
}
type User { ... }
```

**Technical Requirements**
  _Option 1_
  - We can generate `endpoint_id: usize` while creating the blueprint, and use endpoint id to perform a lookup in an Array instead of using a hashmap. This would save us the cost of creating a hash for each upstream call.
  - Generated DLs should be saved in ServerContext.
  
  _Option 2_
  - We could also consider saving the DL in `Expression::Endpoint` with the RequestTemplate.
  - The fact that we use the DL or not should remain with the `Expression::eval`  for now.

**Additional Info:**
- Currently we have one data loader shared across all endpoint. This increases thread contention across HTTP requests. With one DL per endpoint, thread contention will also reduce 🤞 resulting in a little better performance when enabled.

**Limitations:**
- Batching is only supported for GET requests with query params only.

### PR DESCRIPTION
**Pull request by amitksingh1490**
Fixes #448
**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh` to address and fix linting issues.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

### CODE REVIEW

File: src/config/config.rs [hunk 1]
```diff
@@ -165,7 +165,7 @@ pub struct Field {
   pub http: Option<Http>,
   #[serde(rename = "unsafe")]
   pub unsafe_operation: Option<Unsafe>,
-  pub batch: Option<Batch>,
+  pub batch: Option<GroupBy>,
```
Comment (tusharmath) [inline src/config/config.rs:168]:
```suggestion
  pub group_by: Option<GroupBy>,
```

File: src/request_template.rs [hunk 2]
```diff
@@ -17,6 +17,7 @@ pub struct RequestTemplate {
   pub headers: Vec<(String, Mustache)>,
   pub body: Option<Mustache>,
   pub endpoint: Endpoint,
+  pub id: Option<usize>,
```
Comment (tusharmath) [inline src/request_template.rs:20]:
```suggestion
  pub id: usize,
```

File: src/http/server_context.rs [hunk 3]
```diff
@@ -4,23 +4,70 @@ use async_graphql::dataloader::{DataLoader, NoCache};
...
+  for def in blueprint.definitions.iter() {
+    if let Definition::ObjectTypeDefinition(def) = def {
+      for field in &def.fields {
+        if let Some(Expression::Unsafe(Operation::Endpoint(_))) = &field.resolver {
+          let mut data_loader = Arc::new(
+            HttpDataLoader::new(DefaultHttpClient::default(), None)
+              .to_data_loader(server.batch.clone().unwrap_or_default()),
+          );
+          field.directives.iter().for_each(|directive| {
+            if directive.name == "batch" {
```
Comment (tusharmath) [inline src/http/server_context.rs:49]: Batching info can remain inside of the Operation::Endpoint.

File: src/http/server_context.rs [hunk 4]
```diff
@@ -4,23 +4,70 @@ use async_graphql::dataloader::{DataLoader, NoCache};
...
+  blueprint: &Blueprint,
+  server: Server,
+) -> Vec<Arc<DataLoader<HttpDataLoader<DefaultHttpClient>, NoCache>>> {
+  let mut data_loaders = Vec::new();
+  for def in blueprint.definitions.iter() {
+    if let Definition::ObjectTypeDefinition(def) = def {
+      for field in &def.fields {
+        if let Some(Expression::Unsafe(Operation::Endpoint(_))) = &field.resolver {
+          let mut data_loader = Arc::new(
+            HttpDataLoader::new(DefaultHttpClient::default(), None)
```
Comment (tusharmath) [inline src/http/server_context.rs:45]: FIX: client should be passed and not created.

Comment (tusharmath) [reply src/http/server_context.rs:45]: Can enrich the endpoint expression with the DL directly creating a new Blueprint.

File: src/request_template.rs (hunk 2 above)
Comment (tusharmath) [reply src/request_template.rs:20]: id is not required, if we save the DL in the expr.

File: tests/graphql_spec.rs [hunk 5]
```diff
@@ -194,19 +194,15 @@ async fn test_execution() -> std::io::Result<()> {
     let mut config = Config::from_sdl(&spec.server_sdl).unwrap();
     config.server.enable_query_validation = Some(false);
 
-    let blueprint = Blueprint::try_from(&config).unwrap();
-    let schema = blueprint.to_schema(&config.server);
+    let mut blueprint = Blueprint::try_from(&config).unwrap();
+    let server_ctx = ServerContext::new(&mut blueprint, config.server.clone());
+    let schema = server_ctx.schema.clone();
```
[the remark this answers is not in the document]
Comment (amitksingh1490) [reply tests/graphql_spec.rs:199]: will do in separate PR

File: src/blueprint/from_config.rs [hunk 6]
```diff
@@ -313,6 +315,29 @@ fn update_unsafe(field: config::Field, mut b_field: FieldDefinition) -> FieldDef
   b_field
 }
 
+fn update_batch(field: &config::Field, mut b_field: FieldDefinition, _config: &Config) -> Valid<FieldDefinition> {
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:318]:
```suggestion
fn update_group_by(field: &config::Field, mut b_field: FieldDefinition, _config: &Config) -> Valid<FieldDefinition> {
```

File: src/lambda/expression.rs [hunk 7]
```diff
@@ -73,17 +100,16 @@ impl Expression {
...
               // Attempt to short circuit GET request
               if is_get && ctx.req_ctx.server.batch.is_some() {
                 let headers = ctx.req_ctx.server.batch.clone().map(|s| s.headers).unwrap_or_default();
                 let endpoint_key = crate::http::DataLoaderRequest::new(req, headers);
-                let resp = ctx
-                  .req_ctx
-                  .data_loader
+                let resp = dl
                   .as_ref()
+                  .ok_or(EvaluationError::IOException("DataLoader not found".to_string()))?
```
Comment (tusharmath) [inline src/lambda/expression.rs:112]: We can safely unwrap over here. This is an unexpected error.

File: tests/graphql_spec.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tusharmath) [reply tests/graphql_spec.rs:199]: please create a github issue for this.

[the remark this answers is not in the document]
Comment (amitksingh1490) [reply tests/graphql_spec.rs:199]: Here: https://github.com/tailcallhq/tailcall/issues/464

**Review by tusharmath — APPROVED** [review]

**Pull request merged**